types/underscore/underscore-tests.ts

```diff
@@ -726,8 +768,584 @@ interface ChainTypeExtractor {
+    _(intersectingObjectPropertiesDictionary).collect(stringRecordProperty); // $ExpectType (string | boolean)[]
+    extractChainTypes(_.chain(intersectingObjectPropertiesDictionary).collect(stringRecordProperty)); // $ExpectType ChainType<(string | boolean)[], string | boolean>
+
+    // property name iterator with a nullable type union
+    _.map(stringRecordOrUndefinedArray, stringRecordProperty); // $ExpectType (string | undefined)[]
+    _(stringRecordOrUndefinedArray).map(stringRecordProperty); // $ExpectType (string | undefined)[]
+    extractChainTypes(_.chain(stringRecordOrUndefinedArray).map(stringRecordProperty)); // $ExpectType ChainType<(string | undefined)[], string | undefined>
+    _.collect(stringRecordOrUndefinedArray, stringRecordProperty); // $ExpectType (string | undefined)[]
+    _(stringRecordOrUndefinedArray).collect(stringRecordProperty); // $ExpectType (string | undefined)[]
+    extractChainTypes(_.chain(stringRecordOrUndefinedArray).collect(stringRecordProperty)); // $ExpectType ChainType<(string | undefined)[], string | undefined>
+
+    _.map(stringRecordOrUndefinedList, stringRecordProperty); // $ExpectType (string | undefined)[]
+    _(stringRecordOrUndefinedList).map(stringRecordProperty); // $ExpectType (string | undefined)[]
+    extractChainTypes(_.chain(stringRecordOrUndefinedList).map(stringRecordProperty)); // $ExpectType ChainType<(string | undefined)[], string | undefined>
+    _.collect(stringRecordOrUndefinedList, stringRecordProperty); // $ExpectType (string | undefined)[]
+    _(stringRecordOrUndefinedList).collect(stringRecordProperty); // $ExpectType (string | undefined)[]
+    extractChainTypes(_.chain(stringRecordOrUndefinedList).collect(stringRecordProperty)); // $ExpectType ChainType<(string | undefined)[], string | undefined>
+
+    _.map(stringRecordOrUndefinedDictionary, stringRecordProperty); // $ExpectType (string | undefined)[]
+    _(stringRecordOrUndefinedDictionary).map(stringRecordProperty); // $ExpectType (string | undefined)[]
+    extractChainTypes(_.chain(stringRecordOrUndefinedDictionary).map(stringRecordProperty)); // $ExpectType ChainType<(string | undefined)[], string | undefined>
+    _.collect(stringRecordOrUndefinedDictionary, stringRecordProperty); // $ExpectType (string | undefined)[]
+    _(stringRecordOrUndefinedDictionary).collect(stringRecordProperty); // $ExpectType (string | undefined)[]
+    extractChainTypes(_.chain(stringRecordOrUndefinedDictionary).collect(stringRecordProperty)); // $ExpectType ChainType<(string | undefined)[], string | undefined>
+
+    // property name iterator with a non-nullable non-intersecting type union
+    // linting has odd issues correctly determining the UnderscoreStatic.map type results for this case when using ExpectType for some reason
+    // that stop occurring when the minimum version is moved to TS 3.1, so use assignability checks instead for the moment as a workaround
```

Comment:
To provide a little more context around the issues I mention here, linting was initially failing with an expected type of `(string | undefined)[]` and an actual type of `string[]`, but super oddly when I tried switching the expected type to `string[]` as a temporary workaround for these few failures the linter then flip-flopped and failed with an expected type of `string[]` and an actual type of `(string | undefined)[]`.

If there's a different workaround that folks would like me to try or a different course of action that folks would like me to take here, let me know.

Reply:
This is probably one of those cases where TS just isn't doing what it should be doing. The only thing you can do in such a case is a workaround. Workarounds are never clean, but it's all we got.

Reply:
It's true, and at least this one seems to go away in TS 3.1, which I might kick Underscore up to anyway if that's the new minimum version near the end of this set of PRs since `zip` and `unzip` would benefit from that.

Reply:
This ended up going away with the resolution of [this comment chain](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45763#discussion_r448595585).